components/messenger/transports.rst

```diff
@@ -0,0 +1,222 @@
+
+To send and receive messages via message brokers, the Messenger component has
+some transports, to which you can route your message.
+
+Every transport is configurable using a DSN. This DSN allows you to chose the
+transport layer as well as configuring it.
+
+AMQP
+----
+
+Very likely the most famous message broker protocol, AMQP, especially with RabbitMQ
+has a built-in support within the Messenger component.
+
+How it works?
+~~~~~~~~~~~~~
+
+The DSN protocol to use is ``amqp``. The following DSN example shows you how to
+use the adapter to connect to a local RabbitMQ with the ``user`` username and
+``password`` password. The messages will be sent to the ``messages`` exchange
+bound to the ``messages`` queue on the ``/`` vhost::
+
+    amqp://user:password@localhost/%2f/messages
+
+.. note:
+
+    By default, RabbitMQ uses ``guest`` as a username and ``guest`` as a password
+    and has a ``/`` vhost.
+
+Error handling
+~~~~~~~~~~~~~~
+
+If something wrong happens (i.e. an exception is thrown) while handling your message,
+the default behaviour is that your message will be "NACK" and requeued.
+
+However, if your exception implements the ``RejectMessageExceptionInterface`` interface,
+the message will be rejected from the queue.
+
+Retry
+~~~~~
+
+When receiving messages from a broker, it might happen that some exceptions will
+arise. Typically, a 3rd party provider is down or your system is under heavy load
+and can't really process some messages. To handle this scenario, there is a built-in
+retry mechanism that can be enabled via your DSN::
+
+    amqp://guest:guest@localhost/%2f/messages
+      ?retry[attempts]=3
+      &retry[ttl][0]=10000
+      &retry[ttl][1]=30000
+      &retry[ttl][2]=60000
+
+In the example above, if handling the message fails, it will retry it 3 times. After
+the first failure, it will wait 10 seconds before trying it. After the 2nd failure,
+it will wait 30 seconds. After the 3rd failure, it will wait a minute. If it still
+fails, the message will be moved to a "dead queue" and you will have to manually
+handle this message.
+
+DSN configuration reference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The options available to in the DSN are documented on the ``Connection`` class
+in the code repository.
+
+Enqueue
+-------
+
+Probably one of the most famous PHP queue-broker libraries, Enqueue has 10+ adapters
+with brokers like Kafka, Google Pub/Sub, AWS SQS and more. Check out the transport
+documentation in `Enqueue's official repository`_.
+
+Your own transport
+------------------
+
+If there is no available transport for your message broker, you can easily
+create your own.
+
+Your own sender
+~~~~~~~~~~~~~~~
+
+Using the ``SenderInterface``, you can easily create your own message sender.
+Let's say you already have an ``ImportantAction`` message going through the
+message bus and handled by a handler. Now, you also want to send this message as
+an email.
+
+First, create your sender::
+
+    namespace App\MessageSender;
```

Comment:
We could go suggesting a better structure for Messenger stuf? maybe `App\Message\Sender` or `App\Messenger\Transport\Sender` instead?

Reply:
Yep, why not 👍